cypress/plugins/synapsedocker/index.ts

```diff
@@ -0,0 +1,169 @@
+
+            const synapseId = await new Promise<string>((resolve, reject) => {
+                childProcess.execFile('docker', [
+                    "run",
+                    "-d",
+                    "-v", `${synCfg.configDir}:/data`,
+                    "-p", "8008/tcp",
+                    "matrixdotorg/synapse",
+                    "run",
+                ], (err, stdout) => {
+                    if (err) reject(err);
+                    resolve(stdout.trim());
+                });
+            });
+
+            // Get the port that docker allocated: specifying only one
+            // port above leaves docker to just grab a free one, although
+            // in hindsight we need to put the port in public_baseurl in the
+            // config really, so this will probably need changing to use a fixed
+            // / configured port.
+            const port = await new Promise<number>((resolve, reject) => {
+                childProcess.execFile('docker', [
+                    "port", synapseId, "8008",
+                ], { encoding: 'utf8' }, (err, stdout) => {
+                    if (err) reject(err);
+                    resolve(Number(stdout.trim().split(":")[1]));
+                });
+            });
+
+            synapses.set(synapseId, Object.assign({
+                port,
+                synapseId,
+            }, synCfg));
+
+            console.log(`Started synapse with id ${synapseId} on port ${port}.`);
+            return synapses.get(synapseId);
+        },
+        async synapseStop(id) {
+            const synCfg = synapses.get(id);
+
+            if (!synCfg) throw new Error("Unknown synapse ID");
+
+            const dockerLogsPath = path.join("cypress", "dockerlogs", id);
+            await fse.ensureDir(dockerLogsPath);
+
+            await new Promise<void>((resolve, reject) => {
+                childProcess.execFile('docker', [
+                    "logs",
+                    id,
+                ], async (err, stdout, stderr) => {
+                    if (err) reject(err);
+                    await fse.writeFile(path.join(dockerLogsPath, "stdout.log"), stdout);
+                    await fse.writeFile(path.join(dockerLogsPath, "stderr.log"), stderr);
+                    resolve();
+                });
+            });
+
+            await new Promise<void>((resolve, reject) => {
+                childProcess.execFile('docker', [
+                    "stop",
+                    id,
+                ], err => {
+                    if (err) reject(err);
+                    resolve();
+                });
+            });
+
+            await new Promise<void>((resolve, reject) => {
+                childProcess.execFile('docker', [
+                    "rm",
+                    id,
+                ], err => {
+                    if (err) reject(err);
+                    resolve();
+                });
+            });
```

Comment:
this is the code saved by a `--rm` flag on `docker run`

Reply:
Yep, but then we can't get the logs afaics

Reply:
we could add a custom logger, but probably not something for this PR